src/realm/object-store/sync/sync_manager.cpp

```diff
@@ -397,6 +397,25 @@ void SyncManager::close_all_sessions()
     get_sync_client().wait_for_session_terminations();
 }
 
+
+void SyncManager::clear_connection_change_callbacks()
```

Comment:
One idea I had was to expose this through the app and manually run that the way we do `clearAllAppCaches` in this code it isn't yet exposed.

Reply:
have you tried using this (or `close_all_sessions()`)? I am curious if React Native waits for it to complete.

Reply:
I believe I have but will re-check now

Reply:
@danieltabacaru I have actually been running `close_all_sessions` as `App::close_all_sync_sessions` calls this and eventually this hits the `become_dying` > `become_inactive` > `do_become_inactive` function I have been modifying. Without the changes in the PR I don't think the listeners are actually removed.

It actually may not be an awaiting issue since this call gets made in the start of loading the new state after reload rather than the end of the old one so if the function is blocking correctly I believe it should await.

You can see the JS side of things here: https://github.com/realm/realm-js/pull/6824

Reply:
I think I finally understand what's going on. When the app reloads, the connection callback of an old SyncSession is invoked, but the callback is long gone (and hence the crash). Does this sound right?